### PR TITLE
Add `eth-tester` backend

### PR DESCRIPTION
**Pull request by pcppcp**
- adds `tester` blockchain backend that allows test being run with
  `eth-tester`/`py-evm` combo (use `pytest --blockchain-backend=tester` )
- some test fixtures reorganization
- travis tests still use `geth`

check before merging:
- [x] verify https://github.com/ethereum/eth-tester/pull/100 has been merged & published
- [x] replace `eth-tester` url in `requirements.txt` with PyPi library
- [x] remove git link replacement in `setup.py`

[skip ci]

### CODE REVIEW

File: raiden/storage/wal.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
     'InternalEvent',
     ('identifier', 'state_change_id', 'block_number', 'event_object'),
 )
+import structlog
```
[the remark this answers is not in the document]
Comment (pcppcp) [reply raiden/storage/wal.py]: not needed really, I just forgot to remove the import after I've removed logs I had in there.

File: .travis.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ install:
 before_script:
   ./.travis/before_script.sh
 script:
-  - coverage run -m py.test -Wd --travis-fold=always -vvvvvv --log-config='raiden:DEBUG' --random --blockchain-type=geth $TEST
+  - coverage run -m py.test -Wd --travis-fold=always -xvvvvvv --log-config='raiden:DEBUG' --random --blockchain-type=geth $TEST
```
Comment (palango) [inline .travis.yml]: the `-x` should be removed

File: raiden/storage/wal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (palango) [reply raiden/storage/wal.py]: still applies

File: requirements-dev.txt [hunk 3]
```diff
@@ -24,3 +24,7 @@ releases==1.5.0
 
 # Release
 bumpversion==0.5.3
+
+# eth-tester
+git+https://github.com/pcppcp/eth-tester.git@add_status#egg=eth-tester
```
Comment (palango) [inline requirements-dev.txt]: the version with your fixes was released, link to the one on pypi

File: requirements-dev.txt [hunk 4]
```diff
@@ -24,3 +24,7 @@ releases==1.5.0
...
+py-evm==0.2.0a18
```
Comment (palango) [inline requirements-dev.txt]: when using eth-tester[py-evm] this shouldn't be required.

File: raiden/transfer/channel.py [hunk 5]
```diff
@@ -3,6 +3,9 @@
 import heapq
 from binascii import hexlify
 from collections import namedtuple
+import structlog
+
+log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
```
Comment (palango) [inline raiden/transfer/channel.py]: forgotten as well?

File: raiden/transfer/architecture.py [hunk 6]
```diff
@@ -3,6 +3,10 @@
 from copy import deepcopy
 from typing import List
 
+import structlog
+
+log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
+
```
Comment (palango) [inline raiden/transfer/architecture.py]: forgotten as well?

File: requirements-dev.txt (hunk 4 above)
Comment (pcppcp) [reply requirements-dev.txt]: evm 0.2.0a26 wasn't compatible with py-evm, but apparently they've fixed it in the new release... I'll update the requirements.

File: raiden/blockchain/events.py [hunk 7]
```diff
@@ -57,12 +63,16 @@ def get_contract_events(
     `contract_address` that match the filters `topics`, `from_block`, and
     `to_block`.
     """
-    events = chain.client.get_filter_events(
-        contract_address,
-        topics=topics,
-        from_block=from_block,
-        to_block=to_block
-    )
+    try:
```
Comment (palango) [inline raiden/blockchain/events.py]: I think we move this into the client, otherwise we have to repeat it everywhere the `get_filter_events`  method is used.

File: raiden/network/rpc/client.py [hunk 8]
```diff
@@ -186,6 +186,7 @@ def __init__(
             gasprice: int = None,
             nonce_update_interval: float = 5.0,
             nonce_offset: int = 0,
+            web3: Web3 = None
```
Comment (palango) [inline raiden/network/rpc/client.py]: you added a lot of commas, but miss one here

File: raiden/network/rpc/client.py [hunk 9]
```diff
@@ -221,13 +222,20 @@ def __init__(
...
-        self.web3.middleware_stack.inject(geth_poa_middleware, layer=0)
-
-        # create the connection test middleware
-        connection_test = make_connection_test_middleware(self)
-        self.web3.middleware_stack.inject(connection_test, layer=0)
+        if web3 is None:
+            self.web3: Web3 = Web3(HTTPProvider(endpoint))
+        else:
+            self.web3 = web3
+        try:
```
Comment (palango) [inline raiden/network/rpc/client.py]: why is this `try` clause necessary

File: raiden/network/rpc/client.py [hunk 10]
```diff
@@ -221,13 +222,20 @@ def __init__(
...
+            # we use a PoA chain for smoketest, use this middleware to fix this
+            self.web3.middleware_stack.inject(geth_poa_middleware, layer=0)
+        except ValueError:
+            pass
+
+        # create the connection test middleware (but only for non-tester chain)
+        if hasattr(web3, 'testing') is False:
```
Comment (palango) [inline raiden/network/rpc/client.py]: I'd prefer `if not hasattr(web3, 'testing')`

File: raiden/tests/integration/fixtures/backend_tester.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    [ethereum_tester.send_transaction({
+        'from': faucet_address,
+        'to': to_checksum_address(privatekey_to_address(key)),
+        'gas': 21000,
+        'value': 1 * denoms.ether
+    }) for key in private_keys]
+
+
+@pytest.fixture
+def spawn_autominer(web3):
```
Comment (palango) [inline raiden/tests/integration/fixtures/backend_tester.py]: Why is this needed? Is the mining blocking otherwise?

File: raiden/network/rpc/client.py (hunk 9 above)
Comment (pcppcp) [reply raiden/network/rpc/client.py]: Web3 raises a value error if you are trying to inject middleware that already exists (which is what happens with `tester` - `web3` fixture is session scoped)

File: raiden/tests/integration/fixtures/backend_tester.py (hunk 11 above)
Comment (pcppcp) [reply raiden/tests/integration/fixtures/backend_tester.py]: The tester only mines a block once you submit a transaction, but no empty blocks are mined. Block generation is required to confirm the transactions.

File: raiden/blockchain/events.py (hunk 7 above)
Comment (pcppcp) [reply raiden/blockchain/events.py]: ah, you are right, that makes much more sense.

File: raiden/network/rpc/client.py (hunk 8 above)
Comment (pcppcp) [reply raiden/network/rpc/client.py]: I've now added another _large_ bag of commas. All hail the glorious `flake8-commas`.

Comment (palango) [reply raiden/network/rpc/client.py]: Hehe, unfortunately there's no auto-fix with that :(

**Comment by pcppcp** [comment]
@palango addressed your comments and rebased to current `master`

**Comment by LefterisJP** [comment]
@pcppcp some tests have failed in the latest run

**Comment by pcppcp** [comment]
@LefterisJP that's matrix stuff @ulope & @jomuel are looking into it. The CI tests still use `geth` as a backend.

File: raiden/tests/integration/rpc/test_assumptions.py [hunk 12]
```diff
@@ -93,18 +113,18 @@ def test_estimate_gas_fail(deploy_client, blockchain_backend):
...
     """ If the same transaction is sent twice a JSON RPC error is raised. """
     contract_proxy = deploy_rpc_test_contract(deploy_client)
 
     address = contract_proxy.contract_address
     assert len(deploy_client.web3.eth.getCode(to_checksum_address(address))) > 0
 
-    host = '0.0.0.0'  # hardcoded in the deploy_client fixture
     second_client = JSONRPCClient(
-        host,
+        '',
```
Comment (LefterisJP) [inline raiden/tests/integration/rpc/test_assumptions.py]: why did you remove the `'0.0.0.0'` host? Is empty string the same? If yes why remove it and not just be explicit?

File: raiden/tests/unit/conftest.py [hunk 13]
```diff
@@ -0,0 +1,2 @@
+# -*- coding: utf-8 -*-
+# from raiden.tests.integration.fixtures.blockchain import *  # noqa: F401,F403
```
Comment (LefterisJP) [inline raiden/tests/unit/conftest.py]: What's this file for? It just seems to have 1 commented import line and nothing else

File: raiden/tests/utils/blockchain.py [hunk 14]
```diff
@@ -38,6 +38,7 @@ def wait_until_block(chain, block):
     curr_block = chain.block_number()
     while curr_block < block:
         curr_block = chain.next_block()
+        gevent.sleep(0)
```
Comment (LefterisJP) [inline raiden/tests/utils/blockchain.py]: haha ... I used to have this here but removed it when we removed tester the last time. Are we sure that the `0` value allows for a context switch?

Edit: According to the [docs](http://www.gevent.org/api/gevent.html#gevent.sleep)

> In the current implementation, a value of 0 (the default) means to yield execution to any other runnable greenlets, but this greenlet may be scheduled again before the event loop cycles (in an extreme case, a greenlet that repeatedly sleeps with 0 can prevent greenlets that are ready to do I/O from being scheduled for some (small) period of time); a value greater than 0, on the other hand, will delay running this greenlet until the next iteration of the loop.

So perhaps just switch to a very small positive value instead?

File: raiden/transfer/node.py [hunk 15]
```diff
@@ -294,8 +294,7 @@ def subdispatch_targettask(
 
     elif sub_task and isinstance(sub_task, PaymentMappingState.TargetTask):
         is_valid_subtask = (
-            token_network_identifier == sub_task.token_network_identifier and
-            token_address == sub_task.token_address
+            token_network_identifier == sub_task.token_network_identifier
```
Comment (LefterisJP) [inline raiden/transfer/node.py]: Why did you remove the `token_address` check in the state machine here? What does this code have to do with eth-tester?

File: setup.cfg [hunk 16]
```diff
@@ -1,10 +1,10 @@
 [flake8]
-ignore = E731, E402, W504
+ignore = E731, E402, W504, C814
```
Comment (LefterisJP) [inline setup.cfg]: are you use 814 is the right number. In the docs of the flake8 package it [has](https://github.com/PyCQA/flake8-commas#errors) many similar errors listed and the most generic one sounds like 812

File: requirements-dev.txt [hunk 17]
```diff
@@ -25,3 +25,7 @@ releases==1.5.0
 
 # Release
 bumpversion==0.5.3
+
+# eth-tester
```
Comment (LefterisJP) [inline requirements-dev.txt]: do you need the commented out lines?

File: raiden/tests/unit/conftest.py (hunk 13 above)
Comment (pcppcp) [reply raiden/tests/unit/conftest.py]: that's slipped in when rebasing the proxy unit tests. It can go of course.

File: raiden/tests/utils/blockchain.py (hunk 14 above)
Comment (pcppcp) [reply raiden/tests/utils/blockchain.py]: yep, context switch. I was getting infinite loop with tester in there and this seems to help. Good point referring to the docs, I can put something like 1 ms as delay.

File: raiden/transfer/node.py (hunk 15 above)
Comment (pcppcp) [reply raiden/transfer/node.py]: that's a bug in Raiden - `token_address` is no longer part of the `TargetTask` structure, accessing it will cause an exception.

File: requirements-dev.txt (hunk 17 above)
Comment (pcppcp) [reply requirements-dev.txt]: hm, no, I've commented these out when testing dependencies, but since `eth-tester` updated to support most recent `py-evm` the uncommented line is correct and I'll remove the others

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 18]
```diff
@@ -266,6 +277,9 @@ def raiden_network(
     _raiden_cleanup(request, raiden_apps)
 
     # Force blocknumber update
+    exception = RuntimeError('Alarm failed to start and set up start_block correctly')
+    with gevent.Timeout(seconds=5, exception=exception):
+        wait_for_alarm_start(raiden_apps)
```
[the remark this answers is not in the document]
Comment (pcppcp) [reply raiden/tests/integration/fixtures/raiden_network.py]: the problem is Alarm task.
immediately after the alarm is started, it tries to get the block number it was started in
`self.last_block_number = self.chain.block_number()`
But if context switch due to i/o block occurs at this point, `last_block_number` is still `None`
If you take a look at `raiden_network` fixture, it calls `poll_for_new_block()` immediately after the channels are ready. If the alarm is still waiting for block, the `poll_for_new_block()` method will attempt to perform arithmetic with `last_block_number` still set to `None`

File: raiden/tasks.py [hunk 19]
```diff
@@ -77,6 +77,7 @@ def _run(self):  # pylint: disable=method-hidden
 
     def poll_for_new_block(self):
         current_block = self.chain.block_number()
+        log.info('block', block=current_block)
```
[the remark this answers is not in the document]
Comment (pcppcp) [reply raiden/tasks.py]: debug leftovers..

File: raiden/tests/integration/contracts/test_secret_registry.py [hunk 20]
```diff
@@ -5,10 +5,10 @@
 def test_secret_registry(secret_registry_proxy):
     #  register secret
     secret = get_random_bytes(32)
+    event_filter = secret_registry_proxy.secret_registered_filter()
```
[the remark this answers is not in the document]
Comment (pcppcp) [reply raiden/tests/integration/contracts/test_secret_registry.py]: filter failed to log events if it is created after the event occurs. Not sure why.

File: raiden/tests/integration/rpc/test_assumptions.py (hunk 12 above)
Comment (pcppcp) [reply raiden/tests/integration/rpc/test_assumptions.py]: `host` and `port` are not used if `web3` is provided. I was thinking of removing those two variables completely and make `web3` a required argument, but I thought I'd address this in a separate PR as it also affects other parts of the code.

**Comment by pcppcp** [comment]
@LefterisJP addressed your comments with a new commit

> General Question: After this PR is merged we wil also be able to provide --blockchain-type tester for faster tests locally. But in travis it always remains as geth for the moment. Right?

yes, CI tests are still using `geth` as a backend. But we should consider switching at least non-critical integration tests that require blockchain to `tester`. Contract proxies would be first candidate, followed by various transfer tests.

File: raiden/tests/integration/rpc/test_assumptions.py (hunk 12 above)
Comment (LefterisJP) [reply raiden/tests/integration/rpc/test_assumptions.py]: Hmm I am not following.

Web3 is initialized in the ctor of `JSONRPCClient`:

```
        endpoint = 'http://{}:{}'.format(host, port)
        # web3
        self.web3: Web3 = Web3(HTTPProvider(endpoint))
```

File: raiden/transfer/node.py (hunk 15 above)
Comment (LefterisJP) [reply raiden/transfer/node.py]: Hmm I see, yes we missed that one in commit: 0bee56ffca77d72540d808b59806ded434ec013d

The initiator and mediator task had it removed. Can you also remove `token_address` from the function and from where it's called up the chain as it's not needed?

File: raiden/tests/integration/rpc/test_assumptions.py (hunk 12 above)
Comment (pcppcp) [reply raiden/tests/integration/rpc/test_assumptions.py]: you are probably looking at a wrong branch:

https://github.com/pcppcp/raiden/blob/tester_test/raiden/network/rpc/client.py#L231

Comment (LefterisJP) [reply raiden/tests/integration/rpc/test_assumptions.py]: Ah okay indeed. Now I see. So you want to make passing the web3 argument mandatory and then get rid of these args. Cool. This can happen in another PR ... but still in the meantime makes no sense to remove `host` from here in this PR.

File: raiden/tests/integration/rpc/test_assumptions.py [hunk 21]
```diff
@@ -9,6 +9,16 @@
 from raiden.network.rpc.client import JSONRPCClient
 from raiden.utils.solc import compile_files_cwd
 
+try:
```
[the remark this answers is not in the document]
Comment (pcppcp) [reply raiden/tests/integration/rpc/test_assumptions.py]: This is related to how `eth-tester` works - it calls the evm directly on executing contract's `call()`, `estimateGas()` or `transfer()` . If `require` is triggered, it'll raise an exception.
Two exceptions that I am trying to import are raised only when using `eth-tester` as a backend. And because tester is a dev dependency, this makes (and fakes) the exceptions to make python happy in case the library isn't available.
